# Nama: Jade Wulantrisna Aitken
# NIM: 2702279205

### Soal: GANJIL [A]


## Preprocess Data

In [1]:
import pandas as pd
import numpy as np
import pickle as pkl
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt             #visualisation


In [2]:
df= pd.read_csv("Dataset_A_loan.csv")

In [3]:
df.head()

person_age person_gender person_education  person_income  person_emp_exp  \
0        22.0        female           Master        71948.0               0   
1        21.0        female      High School        12282.0               0   
2        25.0        female      High School        12438.0               3   
3        23.0        female         Bachelor        79753.0               0   
4        24.0          male           Master        66135.0               1   

  person_home_ownership  loan_amnt loan_intent  loan_int_rate  \
0                  RENT    35000.0    PERSONAL          16.02   
1                   OWN     1000.0   EDUCATION          11.14   
2              MORTGAGE     5500.0     MEDICAL          12.87   
3                  RENT    35000.0     MEDICAL          15.23   
4                  RENT    35000.0     MEDICAL          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.49                         3.0           561   
1                 0.08                         2.0           504   
2                 0.44                         3.0           635   
3                 0.44                         2.0           675   
4                 0.53                         4.0           586   

  previous_loan_defaults_on_file  loan_status  
0                             No            1  
1                            Yes            0  
2                             No            1  
3                             No            1  
4                             No            1

In [4]:
df.describe()

person_age  person_income  person_emp_exp     loan_amnt  \
count  45000.000000   4.275000e+04    45000.000000  45000.000000   
mean      27.764178   8.018223e+04        5.410333   9583.157556   
std        6.045108   7.331158e+04        6.063532   6314.886691   
min       20.000000   8.000000e+03        0.000000    500.000000   
25%       24.000000   4.726725e+04        1.000000   5000.000000   
50%       26.000000   6.705500e+04        4.000000   8000.000000   
75%       30.000000   9.582125e+04        8.000000  12237.250000   
max      144.000000   5.556399e+06      125.000000  35000.000000   

       loan_int_rate  loan_percent_income  cb_person_cred_hist_length  \
count   45000.000000         45000.000000                45000.000000   
mean       11.006606             0.139725                    5.867489   
std         2.978808             0.087212                    3.879702   
min         5.420000             0.000000                    2.000000   
25%         8.590000             0.070000                    3.000000   
50%        11.010000             0.120000                    4.000000   
75%        12.990000             0.190000                    8.000000   
max        20.000000             0.660000                   30.000000   

       credit_score   loan_status  
count  45000.000000  45000.000000  
mean     632.608756      0.222222  
std       50.435865      0.415744  
min      390.000000      0.000000  
25%      601.000000      0.000000  
50%      640.000000      0.000000  
75%      670.000000      0.000000  
max      850.000000      1.000000

In [5]:
df.shape

(45000, 14)

#### cek apakah ada salah data types. perbaiki.

In [6]:
df.dtypes

person_age                        float64
person_gender                      object
person_education                   object
person_income                     float64
person_emp_exp                      int64
person_home_ownership              object
loan_amnt                         float64
loan_intent                        object
loan_int_rate                     float64
loan_percent_income               float64
cb_person_cred_hist_length        float64
credit_score                        int64
previous_loan_defaults_on_file     object
loan_status                         int64
dtype: object

Data types sudah sesuai semua

#### Rename data for efficiency

In [7]:
df = df.rename(columns={
    "person_age": "age",
    "person_gender": "gender",
    "person_education": "education",
    "person_income": "income",
    "person_emp_exp": "emp_exp",
    "person_home_ownership": "home_own",
    "loan_int_rate": "int_rate",
    "loan_percent_income": "percent_income",
    "cb_person_cred_hist_length": "cred_hist",
    "credit_score": "cred_score",
    "previous_loan_defaults_on_file": "prev_defaults"
})
df.head(5)


age  gender    education   income  emp_exp  home_own  loan_amnt  \
0  22.0  female       Master  71948.0        0      RENT    35000.0   
1  21.0  female  High School  12282.0        0       OWN     1000.0   
2  25.0  female  High School  12438.0        3  MORTGAGE     5500.0   
3  23.0  female     Bachelor  79753.0        0      RENT    35000.0   
4  24.0    male       Master  66135.0        1      RENT    35000.0   

  loan_intent  int_rate  percent_income  cred_hist  cred_score prev_defaults  \
0    PERSONAL     16.02            0.49        3.0         561            No   
1   EDUCATION     11.14            0.08        2.0         504           Yes   
2     MEDICAL     12.87            0.44        3.0         635            No   
3     MEDICAL     15.23            0.44        2.0         675            No   
4     MEDICAL     14.27            0.53        4.0         586            No   

   loan_status  
0            1  
1            0  
2            1  
3            1  
4            1

Categorical data

In [8]:
df['gender'].value_counts()

gender
male       24799
female     20111
Male          45
fe male       45
Name: count, dtype: int64

In [9]:
df['gender'].replace('Male',"male", inplace=True)
df['gender'].replace('fe male',"female", inplace=True)

ubah data yang salah menggunakan .replace

In [10]:
df['education'].value_counts()

education
Bachelor       13399
Associate      12028
High School    11972
Master          6980
Doctorate        621
Name: count, dtype: int64

In [11]:
df['home_own'].value_counts()

home_own
RENT        23443
MORTGAGE    18489
OWN          2951
OTHER         117
Name: count, dtype: int64

In [12]:
df['loan_intent'].value_counts()

loan_intent
EDUCATION            9153
MEDICAL              8548
VENTURE              7819
PERSONAL             7552
DEBTCONSOLIDATION    7145
HOMEIMPROVEMENT      4783
Name: count, dtype: int64

In [13]:
df['prev_defaults'].value_counts()


prev_defaults
Yes    22858
No     22142
Name: count, dtype: int64

Numerical

### SPLIT DATA

In [14]:
X = df.drop(columns=["loan_status"])
y = df["loan_status"]  

# Split dataset (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)


print("Training set size:", X_train.shape)
print("Testing set size:", X_test.shape)

Training set size: (36000, 13)
Testing set size: (9000, 13)


Karena tidak ada ketentuan training dan testing, saya menggunakan ratio paling umum 80:20 


#### handle missing values

In [15]:
X_train.isnull().sum()

age                  0
gender               0
education            0
income            1806
emp_exp              0
home_own             0
loan_amnt            0
loan_intent          0
int_rate             0
percent_income       0
cred_hist            0
cred_score           0
prev_defaults        0
dtype: int64

In [16]:
print('mean income:',np.mean(X_train['income']))

mean income: 80400.59697607768


In [17]:
X_train['income'].fillna(80400.59697607768, inplace=True)
X_test['income'].fillna(80400.59697607768, inplace=True)

#### ENCODING

In [18]:
X_train_GenderEncode={"gender": {"male":1,"female" :0}}

X_train= X_train.replace(X_train_GenderEncode)
X_test= X_test.replace(X_train_GenderEncode)

filename = 'gender_encode.pkl'
pkl.dump(X_train_GenderEncode, open(filename, 'wb'))

In [19]:
X_train_EducationEncode = {"education": {
    "High School": 0,
    "Associate": 1,
    "Bachelor": 2,
    "Master": 3,
    "Doctorate": 4
}}
X_train = X_train.replace(X_train_EducationEncode)
X_test = X_test.replace(X_train_EducationEncode)
pkl.dump(X_train_EducationEncode, open('education_encode.pkl', 'wb'))

In [20]:
X_train_HomeOwnEncode = {"home_own": {"OTHER": 0, "RENT": 1, "MORTGAGE": 2, "OWN": 3}}
X_train = X_train.replace(X_train_HomeOwnEncode)
X_test = X_test.replace(X_train_HomeOwnEncode)
pkl.dump(X_train_HomeOwnEncode, open('home_own_encode.pkl', 'wb'))

In [21]:
X_train_LoanIntentEncode = {"loan_intent": {
    "DEBTCONSOLIDATION": 0,
    "EDUCATION": 1,
    "HOMEIMPROVEMENT": 2,
    "MEDICAL": 3,
    "PERSONAL": 4,
    "VENTURE": 5
}}
X_train = X_train.replace(X_train_LoanIntentEncode)
X_test = X_test.replace(X_train_LoanIntentEncode)
pkl.dump(X_train_LoanIntentEncode, open('loan_intent_encode.pkl', 'wb'))

In [22]:
X_train_PrevDefaultsEncode = {"prev_defaults": {"No": 0, "Yes": 1}}
X_train = X_train.replace(X_train_PrevDefaultsEncode)
X_test = X_test.replace(X_train_PrevDefaultsEncode)
pkl.dump(X_train_PrevDefaultsEncode, open('prev_defaults_encode.pkl', 'wb'))

Masing masing dari data categorical di encode menggunakan binary encoding lalu disimpan ke file pkl untuk digunakan di OOP training 

## Modelling and Training

In [23]:
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

Random Forest

In [24]:
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)
rf_preds = rf_model.predict(X_test)

membuat model random forest dengan random state 42 kemudian di fit dan predict

XGBoost

In [25]:
xgb_model = XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42)
xgb_model.fit(X_train, y_train)
xgb_preds = xgb_model.predict(X_test)

c:\Users\ASUS\anaconda3\Lib\site-packages\xgboost\training.py:183: UserWarning: [22:06:12] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


membuat model xgboost dengan eval metric logloss dan random state 42 kemudian di fit dan predict

## Evaluation

In [26]:
print("Random Forest:\n")

accuracy = accuracy_score(y_test, rf_preds)
print(f"Accuracy: {accuracy:.4f}\n")

print("Classification Report:")
print(classification_report(y_test, rf_preds))

print("Confusion Matrix:")
print(confusion_matrix(y_test, rf_preds))

print("------------------------------------------------------")
print("XGBoost:\n")

accuracy = accuracy_score(y_test, xgb_preds)
print(f"Accuracy: {accuracy:.4f}\n")

print("Classification Report:")
print(classification_report(y_test, xgb_preds))

print("Confusion Matrix:")
print(confusion_matrix(y_test, xgb_preds))


Random Forest:

Accuracy: 0.9280

Classification Report:
              precision    recall  f1-score   support

           0       0.94      0.97      0.95      7000
           1       0.89      0.77      0.83      2000

    accuracy                           0.93      9000
   macro avg       0.91      0.87      0.89      9000
weighted avg       0.93      0.93      0.93      9000

Confusion Matrix:
[[6808  192]
 [ 456 1544]]
------------------------------------------------------
XGBoost:

Accuracy: 0.9377

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.97      0.96      7000
           1       0.90      0.81      0.85      2000

    accuracy                           0.94      9000
   macro avg       0.92      0.89      0.91      9000
weighted avg       0.94      0.94      0.94      9000

Confusion Matrix:
[[6810  190]
 [ 371 1629]]


#### 1. Accuracy
- Random forest -> 92.8%
- XGBoost       -> 93.77%
- XGB > Random Forest
#### 2. Classification Report
- Random Forest -> Class 1       0.89      0.77      0.83      2000
- XGBoost       -> Class 1       0.90      0.81      0.85      2000
- XGB > Random Forest
#### 3. Confusion matriks
- Random Forest -> FN -> 456   FP -> 192
- XGBoost       -> FN -> 371   FP -> 190
- XGB > Random forest 

### KESIMPULAN
- XGBoost adalah pilihan lebih baik untuk deployment

## SAVE MODEL

In [27]:
pkl.dump(xgb_model, open('finalmodel.pkl', 'wb'))

In [40]:
df.describe()

age         income       emp_exp     loan_amnt      int_rate  \
count  34264.000000   34264.000000  34264.000000  34264.000000  34264.000000   
mean      26.410810   70052.937223      4.077311   8275.573138     10.879600   
std        3.866703   31531.809737      3.878335   4782.864862      2.928818   
min       20.000000    8000.000000      0.000000    500.000000      5.420000   
25%       23.000000   46462.250000      1.000000   4800.000000      8.490000   
50%       25.000000   64794.500000      3.000000   7350.000000     11.010000   
75%       29.000000   88576.000000      6.000000  11026.250000     12.910000   
max       39.000000  166681.000000     16.000000  22500.000000     19.690000   

       percent_income     cred_hist    cred_score   loan_status  
count    34264.000000  34264.000000  34264.000000  34264.000000  
mean         0.129714      5.000088    632.236867      0.201436  
std          0.071512      2.651525     47.633773      0.401079  
min          0.010000      2.000000    498.000000      0.000000  
25%          0.070000      3.000000    601.000000      0.000000  
50%          0.120000      4.000000    639.000000      0.000000  
75%          0.180000      7.000000    668.000000      0.000000  
max          0.320000     13.000000    762.000000      1.000000